core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java

```diff
@@ -11,160 +11,249 @@
+   * in [a, a->b, b, b->c].
    *
-   * @param inode the inode to lock
-   * @param parent the expected parent inode
-   * @throws InvalidPathException if the inode is no long consistent with the caller's expectations
+   * @param childName the child to lock
+   * @param mode the mode to lock in
    */
-  public synchronized void lockReadAndCheckParent(InodeView inode, InodeView parent)
-      throws InvalidPathException {
-    inode.lockReadAndCheckParent(parent);
-    mInodes.add(inode);
-    mLockModes.add(LockMode.READ);
+  public void lockEdge(String childName, LockMode mode) {
+    Preconditions.checkState(endsInInode());
+
+    InodeView lastInode = get(numLockedInodes() - 1);
+    Edge edge = new Edge(lastInode.getId(), childName);
+    mEntries.add(Entry.forEdge(mInodeLockManager.lockEdge(edge, mode), edge));
+    mLockMode = mode;
   }
 
   /**
-   * Locks the given inode in read mode, and adds it to this lock list. This method ensures the
-   * parent is the expected parent inode, and the name of the inode is the expected name.
+   * Locks the root edge in the specified mode.
    *
-   * NOTE: This method assumes that the inode path to the parent has been read locked.
+   * @param mode the mode to lock in
+   */
+  public void lockRootEdge(LockMode mode) {
+    Preconditions.checkState(mEntries.isEmpty());
+
+    mEntries.add(Entry.forEdge(mInodeLockManager.lockEdge(ROOT_EDGE, mode), ROOT_EDGE));
+    mLockMode = mode;
+  }
+
+  /**
+   * Leapfrogs the edge write lock forward, reducing the lock list's write-locked scope.
    *
-   * @param inode the inode to lock
-   * @param parent the expected parent inode
-   * @param name the expected name of the inode to be locked
-   * @throws InvalidPathException if the inode is not consistent with the caller's expectations
+   * For example, if the lock list is in write mode with entries [a, a->b, b, b->c],
+   * pushWriteLockedEdge(c, d) will change the list to [a, a->b, b, b->c, c, c->d]. The c->d edge
+   * will be write locked instead of the b->c edge. At least a read lock on b->c will be maintained
+   * throughout the process so that other threads cannot interfere with creates, deletes, or
+   * renames.
+   *
+   * @param inode the inode to add to the lock list
+   * @param pathComponent the child name for the edge to add to the lock list
    */
-  public synchronized void lockReadAndCheckNameAndParent(InodeView inode, InodeView parent,
-      String name) throws InvalidPathException {
-    inode.lockReadAndCheckNameAndParent(parent, name);
-    mInodes.add(inode);
-    mLockModes.add(LockMode.READ);
+  public synchronized void pushWriteLockedEdge(InodeView inode, String pathComponent) {
+    Preconditions.checkState(mLockMode == LockMode.WRITE);
+
+    lockInode(inode, LockMode.READ);
+    lockEdge(pathComponent, LockMode.WRITE);
+    // downgrade the second to last edge lock.
+    downgradeNthToLastEdge(2);
```

Comment:
instead of counting backwards, would it make sense to ensure the last is an edge, get the reference, and then add the new locks, then downgrade the reference we saved earlier? Counting backwards sounds tricky and maybe prone to error.

Reply:
Makes sense, done